index/index.go

```diff
@@ -0,0 +1,393 @@
+					return err
+				}
+			}
+			return nil
+		}
+
+		decryptedID, err := sealedID.Unseal(labelUUID, cryptor)
+		if err != nil {
+			return err
+		}
+
+		// Check if the current Identifier matches the identifier given as input.
+		if decryptedID.Identifier == identifier {
+			// Delete the sealed ID.
+			if err = i.deleteSealedIdentifier(labelUUID); err != nil {
+				return err
+			}
+
+			// Add the next counter and label such that the next label will be updated to current label later.
+			labelUpdates[decryptedID.NextCounter] = labelUUID
+		}
+
+		counter = decryptedID.NextCounter
+	}
+}
+
+// updateLabelOfNext updates the next label (with counter equal to nextCounter) to current label.
+// This operation is necessary after deletions.
+func (i *SecureIndex) updateLabelOfNext(nextCounter uint64, currentLabel uuid.UUID, cryptor crypto.CryptorInterface, tagger crypto.TaggerInterface) error {
+	nextLabelUUID, err := computeLabelUUID(nextCounter, tagger)
+	if err != nil {
+		return err
+	}
+
+	// Get the sealed ID from the IO Provider. If error, the next sealed Identifier does not exist,
+	// and the function can terminate.
+	nextSealedID, err := i.getSealedIdentifier(nextLabelUUID)
+	if err != nil {
+		return nil
+	}
+
+	// As the label of the next sealed ID should be updated (and not the sealed ID itself),
+	// it is necessary to delete it and add it with the new label.
+	if err = i.deleteSealedIdentifier(nextLabelUUID); err != nil {
+		return err
+	}
+
+	nextDecryptedID, err := nextSealedID.Unseal(nextLabelUUID, cryptor)
+	if err != nil {
+		return err
+	}
+
+	*nextSealedID, err = nextDecryptedID.Seal(currentLabel, cryptor)
+	if err != nil {
+		return err
+	}
+
+	if err = i.putSealedIdentifier(currentLabel, nextSealedID); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Compute the current last Identifier containing the given keyword, i.e. the current Identifier with
+// the largest value of counter.
+func (i *SecureIndex) lastIdentifier(keyword string) (data.Identifier, error) {
+	tagger, cryptor, err := i.getTaggerAndCryptor(keyword)
+	if err != nil {
+		return data.Identifier{}, err
+	}
+
+	// Starting with counter = 0, check if the corresponding keyword label exists in the secure index.
+	// As long as the keyword label exists, update counter to the next counter.
+	counter := uint64(0)
+
+	decryptedID := data.Identifier{}
+
+	for {
+		labelUUID, err := computeLabelUUID(counter, tagger)
+		if err != nil {
+			return data.Identifier{}, err
+		}
+
+		// Get the sealed Identifier from the IO Provider. If error, then the last Identifier has already
+		// been found, and the function can return it.
+		sealedID, err := i.getSealedIdentifier(labelUUID)
+		if err != nil {
+			return decryptedID, nil
+		}
+
+		decrypted, err := sealedID.Unseal(labelUUID, cryptor)
+		if err != nil {
+			return data.Identifier{}, err
+		}
+
+		counter = decryptedID.NextCounter
+		decryptedID = decrypted
+	}
```

Comment:
You can simplify this a bit by exploiting the zero value of `Identifier.NextCounter`. Note that a similar change can be made in `Search` and `Delete`: 
```suggestion
	decryptedID := data.Identifier{}

	for {
		labelUUID, err := computeLabelUUID(decryptedID.NextCounter, tagger)
		if err != nil {
			return data.Identifier{}, err
		}

		// Get the sealed Identifier from the IO Provider. If error, then the last Identifier has already
		// been found, and the function can return it.
		sealedID, err := i.getSealedIdentifier(labelUUID)
		if err != nil {
			return decryptedID, nil
		}

		decryptedID, err = sealedID.Unseal(labelUUID, cryptor)
		if err != nil {
			return data.Identifier{}, err
		}
	}
```